### ISSUE TITLE
v0.25 breaks my test cases - totally different training losses than v0.24

### ISSUE THREAD

**Issue by danielhanchen** [huggingface/accelerate#2208]
### System Info

```Shell
- `Accelerate` version: 0.24.1
- Platform: Linux-5.15.120+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Numpy version: 1.23.5
- PyTorch version (GPU?): 2.1.0+cu118 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 51.00 GB
- GPU type: Tesla T4
- `Accelerate` default config:
	Not found
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Accelerate v0.24.1: https://colab.research.google.com/drive/1VD4PXDhcx1t5dM2hYewyFK891GNfcELI?usp=sharing

Accelerate v0.25.0: https://colab.research.google.com/drive/1-5MdIOp0cM0scC-CdRZhh8OYhnGHqct4?usp=sharing

### Expected behavior

During testing for https://github.com/unslothai/unsloth, the first 7 steps for training errors on v0.24.1 should be:

|Step|Loss|
|-----|-----|
1 | 1.051200
2 | 1.317400
3 | 1.874600
4 | 1.451000
5 | 1.611200
6 | 1.385900
7 | 1.443000

But then on v0.25:
|Step|Loss|
|-----|-----|
1 | 1.691600
2 | 1.662400
3 | 1.790300
4 | 1.853500
5 | 1.421400
6 | 1.478500
7 | 1.437300

The Colab notebook's seed is 3407. I tried it on Tesla T4, A100 and the different training losses persist.

I can't really theorize exactly what's going on, but I narrowed it down to accelerate's new version.

1. I checked if bitsandbytes, transformers, datasets, peft, trl and even tim dettmer's dataset whether they changed - all no
2. Accelerate got updated 20 hrs ago - and the Colab notebooks show if you install an old version, the errors are again equivalent.

**Comment by muellerzr** [comment]
What if you used <0.24.0? We hit a bug where the data was never being shuffled so you were overfitting on the same samples being shown in the same order each time

**Comment by danielhanchen** [comment]
@muellerzr Let me run it now!

**Comment by muellerzr** [comment]
Then that’s how it should be and things are running properly.

**Comment by danielhanchen** [comment]
@muellerzr waitt so you mean v 0.22, 0.23, 0.24 were all incorrectly handling the sampling? Oh my! Good catch with the find!!!

**Comment by muellerzr** [comment]
No, other way around. They *are* handling it right. 0.24 was *not*

**Comment by muellerzr** [comment]
I can look at disabling it if needed, but the point of it is so you can set the seed/randomness every epoch, similar to what the Trainer did/does

**Comment by danielhanchen** [comment]
@muellerzr Oh I'm not too sure sorry - what do you think is the best approach?

I normally manually check if all errors match via Google sheets, so I'm not too sure if other people might get the same problem when they upgrade to v0.25.

But all in all - if it's just the random state changing, I doubt it's problematic - it just means test cases which relied on `torch.allclose` etc will need updating

**Comment by muellerzr** [comment]
If you notice a performance *degradation* that's not good. In general I saw performance improvements when doing my tests, for instance here are the results from the `nlp_example` script on a single GPU from all the versions (read as accuracy | F1 and number is epoch):

```
0.25.0:
1. 0.81, 0.87
2. 0.867, 0.9
3. 0.8823, 0.917
0.24.1:
1. 0.78, 0.86
2. 0.825, 0.866
3. 0.843, 0.889
0.23.0:
1. 0.78, 0.86
2. 0.825, 0.866
3. 0.843, 0.889
```

What I may do is finally create a `DataLoaderConfiguration` which can disable the per-epoch-randomness if plugged in.

**Comment by danielhanchen** [comment]
@muellerzr Ohh noo I doubt it's a performance degradation - you're comparing final epoch error, whilst I'm comparing per step error.

If in fact you show that at the end of each epoch the F1 accuracy is the same or better, I would trust your changes and just update my comparisons :)

I normally compare each step since after every change I do to make Llama training faster, I have to verify if the errors match - but I don't want to wait for 5 hours to get a final error - I'll just compare the first 60 steps for eg.

**Comment by muellerzr** [comment]
No worries at all! I also don't like that we radically shifted the performance of the library/framework though. It should be configurable in the end. If possible though just so we can make sure, if you'd be willing to run your script again and verify the performance does increase on your seed, that would be great. If it is a degradation instead, all the more reason. (And real data is better than something like our little NLP example :) )

**Comment by danielhanchen** [comment]
@muellerzr I'll do just that! I'll report back to compare the entire training loss curves!

**Comment by Qubitium** [comment]
We also oberserved huge diff in train/loss between 0.24 and 0.25

lama2 13B native bf16 finetune using sft and flashattn2 with all train params being equal other than accelerate pkg:
```
0.25 train loss: 0.5519
0.24 train loss: 0.4090
```

**Comment by muellerzr** [comment]
@Qubitium is this the final training loss here? If so then I'll go ahead and make it configurable.

**Comment by Qubitium** [comment]
Yes, this is the final train loss on 1 gpu over 6 epochs.

**Comment by muellerzr** [comment]
Thanks @Qubitium.

Can either of you run your experiment again using accelerate installed via `pip install git+https://github.com/huggingface/accelerate@disable-seedale-rs`? This should bring the old behavior back

**Comment by danielhanchen** [comment]
Sorry @muellerzr didn't report back on the final loss - but to insert myself in the convo with @Qubitium and you, since @Qubitium tested over epochs, and the old behaivor was not randomizing across epochs (if I'm not mistaken), could that be the reason possibly why 0.24's train loss is lower than 0.25?

I'm not sure on eval loss though. Sadly I don't have a beefy GPU so running eval will be quite painful.

I can test the new PR as well!

**Comment by muellerzr** [comment]
Hi @danielhanchen @Qubitium, after some thorough testing we've found that *if you set the seed using `accelerate.utils.set_seed`*, the results should be negligible. I saw a difference of ~0.02% accuracy when training from scratch on the cv example. However, since this is needed we're going to revert this behavior and keep the old one as a default, but this is just a different sampling technique at the end of the day with a pinch more setup.

**Comment by danielhanchen** [comment]
@muellerzr Oh ok interesting - sorry didn't get to testing it (kinda forgot my my end so much apologies!!) Happy New Year!
